src/rewrite_gelu.cpp

```diff
@@ -53,7 +58,60 @@ struct find_gelu_erf
     }
 };
 
-void rewrite_gelu::apply(module& m) const { match::find_matches(m, find_gelu_erf{}); }
+/**
+ * Find fastGELU blocks (where the graph already does a GELU approximation) and replace them
+ * with an alternative approximation that is less likely to overflow.
+ * The replacement approximation is equivalent to:
+ * GELU(x) ~= 0.5 * x * ( 1 + tanh( sqrt(2/M_PI) * (x + 0.044715 * x^3)))
+ */
+struct find_tanh_fast_gelu
+{
+    auto matcher() const { return match::gelu_tanh(); }
+
+    void apply(module& m, const match::matcher_result& r) const
+    {
+        // Algebraically rearranged version of the tanh GELU approximation that is more numerically
+        // stable
+        auto ins        = r.result;
+        auto x          = r.instructions["x"];
+        auto sqrt_2_rpi = m.add_literal(
+            literal{shape{x->get_shape().type()},
+                    {0.7978845608028653558798921198687637369517172623298693153318516593}});
+        auto fit_const = m.add_literal(literal{shape{x->get_shape().type()}, {0.044715f}});
+        auto one       = m.add_literal(literal{shape{x->get_shape().type()}, {1.0f}});
+        auto xb        = insert_common_op(m, ins, make_op("mul"), {x, sqrt_2_rpi});
+        auto a         = insert_common_op(m, ins, make_op("mul"), {xb, fit_const});
+        auto b         = m.insert_instruction(ins, make_op("mul"), a, x);
+        auto c         = m.insert_instruction(ins, make_op("mul"), b, x);
+        auto u         = m.insert_instruction(ins, make_op("add"), c, xb);
+        auto neg_u     = m.insert_instruction(ins, make_op("neg"), u);
+        auto d         = m.insert_instruction(ins, make_op("sub"), neg_u, u);
+        auto emu       = m.insert_instruction(ins, make_op("exp"), d);
+        auto e         = insert_common_op(m, ins, make_op("add"), {one, emu});
+        auto cdf       = insert_common_op(m, ins, make_op("div"), {one, e});
+        auto y         = m.insert_instruction(ins, make_op("mul"), x, cdf);
+        m.replace_instruction(ins, y);
+
+        /* erf() version of GELU
+        auto ins      = r.result;
+        auto x        = r.instructions["x"];
+        auto sqrt1_2  = m.add_literal(literal{shape{x->get_shape().type()}, {M_SQRT1_2}});
+        auto one      = m.add_literal(literal{shape{x->get_shape().type()}, {1.0f}});
+        auto one_half = m.add_literal(literal{shape{x->get_shape().type()}, {0.5f}});
+        auto a        = insert_common_op(m, ins, make_op("mul"), {x, sqrt1_2});
+        auto erf      = m.insert_instruction(ins, make_op("erf"), a);
+        auto add_erf  = insert_common_op(m, ins, make_op("add"), {one, erf});
+        auto b        = insert_common_op(m, ins, make_op("mul"), {one_half, add_erf});
+        auto y        = m.insert_instruction(ins, make_op("mul"), x, b);
+        m.replace_instruction(ins, y);
+        */
+    }
+};
+
+void rewrite_gelu::apply(module& m) const
+{
+    match::find_matches(m, find_gelu_erf{}, find_tanh_fast_gelu{});
```

Comment:
`find_gelu_erf` should only be enabled when fast math is enabled. `find_tanh_fast_gelu` should always be used.